File: scripts/eval/summarize-arenas.py
#!/usr/bin/env python3
"""summarize-arenas.py — three-bucket capability table from external-arenas.sh output.

Reads the per-arena `*.results.json` sidecars written by harness-v0-neural and
prints the neural-only / both-pass / v0-only / both-fail buckets per arena. For
the postal-standards arena it also joins back to the source JSONL (on `input`)
to break the buckets down by edge_class — the dimension the harness sidecar
drops.

Usage: summarize-arenas.py <out-dir> <postal-cases.jsonl>
"""
import json
import sys
from collections import defaultdict


def buckets(results):
    n = len(results)
    both = sum(1 for r in results if r["v0_pass"] and r["neural_pass"])
    only_n = sum(1 for r in results if not r["v0_pass"] and r["neural_pass"])
    only_v0 = sum(1 for r in results if r["v0_pass"] and not r["neural_pass"])
    neither = sum(1 for r in results if not r["v0_pass"] and not r["neural_pass"])
    tree_ok = sum(1 for r in results if r.get("neural_tree_valid"))
    return n, both, only_n, only_v0, neither, tree_ok


def pct(x, n):
    return f"{100*x/n:.0f}%" if n else "—"


def main():
    out_dir, postal_src = sys.argv[1], sys.argv[2]
    arenas = ["libpostal", "perturb", "postal"]

    print("| arena | n | v0 | neural | both | neural-only | v0-only | both-fail | tree-valid |")
    print("| --- | --: | --: | --: | --: | --: | --: | --: | --: |")
    loaded = {}
    for a in arenas:
        try:
            res = json.load(open(f"{out_dir}/{a}.results.json"))
        except FileNotFoundError:
            print(f"| {a} | (no results) |")
            continue
        loaded[a] = res
        n, both, only_n, only_v0, neither, tree_ok = buckets(res)
        v0 = sum(1 for r in res if r["v0_pass"])
        ne = sum(1 for r in res if r["neural_pass"])
        print(f"| {a} | {n} | {pct(v0,n)} | {pct(ne,n)} | {pct(both,n)} "
              f"| {pct(only_n,n)} | {pct(only_v0,n)} | {pct(neither,n)} | {pct(tree_ok,n)} |")

    # postal edge-class breakdown (join on input)
    if "postal" in loaded:
        ec = {}
        for line in open(postal_src):
            row = json.loads(line)
            ec[row["input"]] = row.get("edge_class", "?")
        by = defaultdict(list)
        for r in loaded["postal"]:
            by[ec.get(r["input"], "?")].append(r)
        print("\n### postal arena by edge_class")
        print("| edge_class | n | v0 | neural | both | neural-only | v0-only |")
        print("| --- | --: | --: | --: | --: | --: | --: |")
        for cls in sorted(by):
            res = by[cls]
            n, both, only_n, only_v0, _, _ = buckets(res)
            v0 = sum(1 for r in res if r["v0_pass"])
            ne = sum(1 for r in res if r["neural_pass"])
            print(f"| {cls} | {n} | {pct(v0,n)} | {pct(ne,n)} | {pct(both,n)} "
                  f"| {pct(only_n,n)} | {pct(only_v0,n)} |")


if __name__ == "__main__":
    main()

File: scripts/eval/summarize-arenas.ts
/**
 * @copyright Sister Software
 * @license AGPL-3.0
 * @author Teffen Ellis, et al.
 *
 *   Summarize-arenas.ts — per-arena pass-rate table from external-arenas.ts output.
 *
 *   Reads the per-arena `*.results.json` sidecars written by harness-neural. For the
 *   postal-standards arena it also joins back to the source JSONL (on `input`) to break the rates
 *   down by edge_class — the dimension the harness sidecar drops.
 *
 *   Ported faithfully from scripts/eval/summarize-arenas.py (pure JSON, no numpy); the v0 buckets
 *   went with the rules parser (v7 excision #1151 — harness-v0-neural was deleted with it).
 *
 *   Usage: node scripts/eval/summarize-arenas.ts <out-dir>
 *   <postal-cases.jsonl>
 */

import { readFileSync } from "node:fs"
import { parseArgs } from "node:util"

import { pyFixed } from "@mailwoman/core/utils"

const { positionals } = parseArgs({ allowPositionals: true, strict: false })

interface Result {
	neural_pass: boolean
	neural_tree_valid?: boolean
	input: string
}

function pct(x: number, n: number): string {
	return n ? `${pyFixed((100 * x) / n, 0)}%` : "—"
}

function main(): void {
	const [outDir, postalSrc] = [positionals[0]!, positionals[1]!]
	const arenas = ["libpostal", "perturb", "postal"]

	console.log("| arena | n | neural | fail | tree-valid |")
	console.log("| --- | --: | --: | --: | --: |")

	const loaded: Record<string, Result[]> = {}

	for (const a of arenas) {
		let res: Result[]

		try {
			res = JSON.parse(readFileSync(`${outDir}/${a}.results.json`, "utf8"))
		} catch (error) {
			if ((error as NodeJS.ErrnoException).code === "ENOENT") {
				console.log(`| ${a} | (no results) |`)

				continue
			}

			throw error
		}

		loaded[a] = res
		const n = res.length
		const ne = res.filter((r) => r.neural_pass).length
		const treeOk = res.filter((r) => r.neural_tree_valid).length

		console.log(`| ${a} | ${n} | ${pct(ne, n)} | ${pct(n - ne, n)} | ${pct(treeOk, n)} |`)
	}

	// postal edge-class breakdown (join on input)
	if ("postal" in loaded) {
		const ec: Record<string, string> = {}

		for (const line of readFileSync(postalSrc, "utf8").split("\n")) {
			if (!line) continue
			const row = JSON.parse(line)
			ec[row.input] = row.edge_class ?? "?"
		}

		const by: Record<string, Result[]> = {}

		for (const r of loaded.postal!) {
			const cls = ec[r.input] ?? "?"
			;(by[cls] ??= []).push(r)
		}

		console.log("\n### postal arena by edge_class")
		console.log("| edge_class | n | neural | fail |")
		console.log("| --- | --: | --: | --: |")

		for (const cls of Object.keys(by).toSorted()) {
			const res = by[cls]!
			const n = res.length
			const ne = res.filter((r) => r.neural_pass).length

			console.log(`| ${cls} | ${n} | ${pct(ne, n)} | ${pct(n - ne, n)} |`)
		}
	}
}

if (import.meta.main) {
	main()
}
